stan/math/rev/core/vari.hpp

```diff
@@ -4,12 +4,53 @@
 #include <stan/math/rev/core/chainable_alloc.hpp>
 #include <stan/math/rev/core/chainablestack.hpp>
 #include <ostream>
+#include <type_traits>
 
 namespace stan {
 namespace math {
 
 // forward declaration of var
-class var;
+template <typename T>
+class var_value;
+
+/**
+ * Pure virtual class that all `vari_value` and it's derived classes inherit.
```

Comment:
```suggestion
 * Abstract class that all `vari_value` and it's derived classes inherit.
```

AFAIK in C++ terminology "pure virtual" applies to functions. A class that contains at least one pure virtual function is called abstract class.